# WEATHER FORECASTING

## My First Project

In [188]:
import pandas as pd # data handling
import matplotlib.pyplot as plt # visualization
import seaborn as sns # visualization

- pandas is used for data manipulation

### Loading Datasets

In [189]:
df = pd.read_csv("weather.csv")
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


In [190]:
df.head()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.4,18.4,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55


In [191]:
df.tail()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9
136632,Zimbabwe,Harare,-17.8178,31.0447,Africa/Harare,1776666600,2026-04-20 08:30,20.0,68.1,Sunny,...,7.85,8.05,1,1,06:06,17:43,09:07,20:18,Waxing Crescent,9


In [192]:
df.sample(6)

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
125907,Belgium,Brussels,50.8333,4.3333,Europe/Brussels,1771915500,2026-02-24 07:45,11.0,51.8,Light rain,...,8.95,12.65,1,1,07:38,18:15,10:02,02:30,First Quarter,44
51446,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1738751400,2025-02-05 17:30,19.4,66.9,Overcast,...,120.99,123.21,4,10,06:32,17:50,11:07,No moonset,First Quarter,43
44268,Vanuatu,Port Vila,-17.7333,168.3167,Pacific/Efate,1735554600,2024-12-30 21:30,25.2,77.4,Partly cloudy,...,15.91,24.05,2,2,05:14,18:24,04:05,17:55,Waning Crescent,2
1478,Mozambique,Maputo,-25.9700,32.5900,Africa/Maputo,1716387300,2024-05-22 16:15,26.0,78.8,Partly cloudy,...,4.00,22.30,1,1,06:24,17:08,16:13,05:12,Waxing Gibbous,97
115550,Trinidad and Tobago,Port Of Spain,10.6500,-61.5167,America/Port_of_Spain,1767164400,2025-12-31 03:00,25.2,77.4,Partly Cloudy,...,6.75,10.25,1,1,06:24,17:55,15:06,03:16,Waxing Gibbous,85
129987,Turkey,Ankara,39.9272,32.8644,Europe/Istanbul,1773729900,2026-03-17 09:45,3.3,37.9,Sunny,...,52.35,74.75,3,6,06:57,18:57,06:06,17:19,Waning Crescent,6


In [193]:
df.dtypes # shows column heading on left and data types on right

country                             str
location_name                       str
latitude                        float64
longitude                       float64
timezone                            str
last_updated_epoch                int64
last_updated                        str
temperature_celsius             float64
temperature_fahrenheit          float64
condition_text                      str
wind_mph                        float64
wind_kph                        float64
wind_degree                       int64
wind_direction                      str
pressure_mb                       int64
pressure_in                     float64
precip_mm                       float64
precip_in                       float64
humidity                          int64
cloud                             int64
feels_like_celsius              float64
feels_like_fahrenheit           float64
visibility_km                   float64
visibility_miles                  int64
uv_index                        float64


In [194]:
df.columns # display column names

Index(['country', 'location_name', 'latitude', 'longitude', 'timezone',
       'last_updated_epoch', 'last_updated', 'temperature_celsius',
       'temperature_fahrenheit', 'condition_text', 'wind_mph', 'wind_kph',
       'wind_degree', 'wind_direction', 'pressure_mb', 'pressure_in',
       'precip_mm', 'precip_in', 'humidity', 'cloud', 'feels_like_celsius',
       'feels_like_fahrenheit', 'visibility_km', 'visibility_miles',
       'uv_index', 'gust_mph', 'gust_kph', 'air_quality_Carbon_Monoxide',
       'air_quality_Ozone', 'air_quality_Nitrogen_dioxide',
       'air_quality_Sulphur_dioxide', 'air_quality_PM2.5', 'air_quality_PM10',
       'air_quality_us-epa-index', 'air_quality_gb-defra-index', 'sunrise',
       'sunset', 'moonrise', 'moonset', 'moon_phase', 'moon_illumination'],
      dtype='str')

In [195]:
df.shape # display no. of rows and columns

(136633, 41)

In [196]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 41 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   country                       136633 non-null  str    
 1   location_name                 136633 non-null  str    
 2   latitude                      136633 non-null  float64
 3   longitude                     136633 non-null  float64
 4   timezone                      136633 non-null  str    
 5   last_updated_epoch            136633 non-null  int64  
 6   last_updated                  136633 non-null  str    
 7   temperature_celsius           136633 non-null  float64
 8   temperature_fahrenheit        136633 non-null  float64
 9   condition_text                136633 non-null  str    
 10  wind_mph                      136633 non-null  float64
 11  wind_kph                      136633 non-null  float64
 12  wind_degree                   136633 non-null  int64  


- There are total 136633 entries
- There are 41 columns
- Data types used : 21 float64, 9 int64 and 11 str
- 42.7 MB memory is used
- There are no null columns
- Some column's data type have mistakes (time should have int data type)

In [197]:
df.latitude

0         34.5200
1         41.3300
2         36.7600
3         42.5000
4         -8.8400
           ...   
136628    10.5000
136629    21.0333
136630    15.3547
136631   -15.4167
136632   -17.8178
Name: latitude, Length: 136633, dtype: float64

In [198]:
df[['sunrise','moonset']]

,sunrise,moonset
0,04:50,01:11
1,05:21,02:14
2,05:40,02:14
3,06:31,03:31
4,06:12,00:38
...,...,...
136628,06:14,22:12
136629,05:34,21:34
136630,05:46,21:39
136631,06:15,20:35


### Filtering

In [199]:
df[df['latitude']<30]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
5,Antigua and Barbuda,Saint John's,17.1200,-61.8500,America/Antigua,1715849100,2024-05-16 04:45,26.0,78.8,Partly cloudy,...,1.20,4.50,1,1,05:36,18:32,13:05,01:14,Waxing Gibbous,55
6,Argentina,Buenos Aires,-34.5900,-58.6700,America/Argentina/Buenos_Aires,1715849100,2024-05-16 05:45,8.0,46.4,Clear,...,4.00,5.30,1,1,07:43,17:59,14:36,01:04,Waxing Gibbous,55
8,Australia,Canberra,-35.2800,149.2200,Australia/Sydney,1715849100,2024-05-16 18:45,9.0,48.2,Clear,...,3.70,5.40,1,1,06:52,17:07,13:31,No moonset,Waxing Gibbous,55
11,Bahamas,Nassau,25.0800,-77.3500,America/Nassau,1715849100,2024-05-16 04:45,27.0,80.6,Partly cloudy,...,12.20,24.60,1,2,06:24,19:48,14:01,02:26,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


In [200]:
df[df['longitude']>30]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
7,Armenia,Yerevan,40.1800,44.5100,Asia/Yerevan,1715849100,2024-05-16 12:45,19.0,66.2,Partly cloudy,...,0.80,0.90,1,1,05:45,20:12,13:17,02:31,Waxing Gibbous,55
8,Australia,Canberra,-35.2800,149.2200,Australia/Sydney,1715849100,2024-05-16 18:45,9.0,48.2,Clear,...,3.70,5.40,1,1,06:52,17:07,13:31,No moonset,Waxing Gibbous,55
10,Azerbaijan,Baku,40.4000,49.8800,Asia/Baku,1715849100,2024-05-16 12:45,17.0,62.6,Partly cloudy,...,1.90,2.20,1,1,05:23,19:51,12:54,02:10,Waxing Gibbous,55
12,Bahrain,Manama,26.2400,50.5800,Asia/Bahrain,1715849100,2024-05-16 11:45,36.0,96.8,Sunny,...,39.00,114.30,2,4,04:50,18:18,12:10,00:45,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136626,Uzbekistan,Tashkent,41.3167,69.2500,Asia/Tashkent,1776666600,2026-04-20 11:30,22.9,73.3,Partly Cloudy,...,11.45,13.65,1,1,05:37,19:08,07:01,23:14,Waxing Crescent,8
136627,Vanuatu,Port Vila,-17.7333,168.3167,Pacific/Efate,1776666600,2026-04-20 17:30,26.2,79.2,Overcast,...,6.45,9.65,1,1,05:57,17:35,08:31,19:46,Waxing Crescent,6
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8


In [201]:
df[(df['temperature_celsius']<20)&(df['air_quality_us-epa-index']==1)]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
6,Argentina,Buenos Aires,-34.5900,-58.6700,America/Argentina/Buenos_Aires,1715849100,2024-05-16 05:45,8.0,46.4,Clear,...,4.00,5.30,1,1,07:43,17:59,14:36,01:04,Waxing Gibbous,55
7,Armenia,Yerevan,40.1800,44.5100,Asia/Yerevan,1715849100,2024-05-16 12:45,19.0,66.2,Partly cloudy,...,0.80,0.90,1,1,05:45,20:12,13:17,02:31,Waxing Gibbous,55
8,Australia,Canberra,-35.2800,149.2200,Australia/Sydney,1715849100,2024-05-16 18:45,9.0,48.2,Clear,...,3.70,5.40,1,1,06:52,17:07,13:31,No moonset,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136613,Bulgaria,Lom,43.8139,23.2361,Europe/Sofia,1776666600,2026-04-20 09:30,15.0,59.0,Sunny,...,13.35,14.35,1,2,06:36,20:16,08:00,No moonset,Waxing Crescent,8
136621,Ukraine,Kyiv,50.4333,30.5167,Europe/Kyiv,1776666600,2026-04-20 09:30,5.9,42.6,Cloudy,...,9.25,13.45,1,1,05:55,20:00,06:55,No moonset,Waxing Crescent,8
136623,United Kingdom,London,51.5171,-0.1062,Europe/London,1776666600,2026-04-20 07:30,3.2,37.8,Sunny,...,14.35,18.85,1,2,05:54,20:05,06:54,No moonset,Waxing Crescent,9
136624,United States of America,Washington Harbor,48.0792,-123.0439,America/Los_Angeles,1776666600,2026-04-19 23:30,13.5,56.3,Clear,...,5.45,9.65,1,1,06:13,20:10,07:44,00:01,Waxing Crescent,11


In [202]:
df.temperature_fahrenheit

0         79.8
1         66.2
2         73.4
3         43.3
4         78.8
          ... 
136628    63.6
136629    82.9
136630    69.7
136631    70.8
136632    68.1
Name: temperature_fahrenheit, Length: 136633, dtype: float64

In [203]:
# selecting columns
df['condition_text']

0         Partly Cloudy
1         Partly cloudy
2                 Sunny
3         Light drizzle
4         Partly cloudy
              ...      
136628    Partly Cloudy
136629           Cloudy
136630            Sunny
136631            Sunny
136632            Sunny
Name: condition_text, Length: 136633, dtype: str

In [204]:
# selecting rows i=index and loc = location
df.iloc[19] # single row


country                                                         Bhutan
location_name                                                  Thimphu
latitude                                                         27.48
longitude                                                         89.6
timezone                                                  Asia/Thimphu
last_updated_epoch                                          1715849100
last_updated                                          2024-05-16 14:45
temperature_celsius                                               18.2
temperature_fahrenheit                                            64.8
condition_text                  Patchy light rain in area with thunder
wind_mph                                                           7.2
wind_kph                                                          11.5
wind_degree                                                        142
wind_direction                                                      SE
pressu

In [205]:
df.iloc[ 3:64] # portion

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
5,Antigua and Barbuda,Saint John's,17.12,-61.85,America/Antigua,1715849100,2024-05-16 04:45,26.0,78.8,Partly cloudy,...,1.2,4.5,1,1,05:36,18:32,13:05,01:14,Waxing Gibbous,55
6,Argentina,Buenos Aires,-34.59,-58.67,America/Argentina/Buenos_Aires,1715849100,2024-05-16 05:45,8.0,46.4,Clear,...,4.0,5.3,1,1,07:43,17:59,14:36,01:04,Waxing Gibbous,55
7,Armenia,Yerevan,40.18,44.51,Asia/Yerevan,1715849100,2024-05-16 12:45,19.0,66.2,Partly cloudy,...,0.8,0.9,1,1,05:45,20:12,13:17,02:31,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
59,Finland,Helsinki,60.18,24.93,Europe/Helsinki,1715849100,2024-05-16 11:45,22.0,71.6,Sunny,...,5.5,6.8,1,1,04:38,21:57,12:50,03:47,Waxing Gibbous,55
60,France,Paris,48.87,2.33,Europe/Paris,1715849100,2024-05-16 10:45,15.0,59.0,Partly cloudy,...,9.8,13.6,1,1,06:07,21:28,13:57,03:41,Waxing Gibbous,55
61,Gabon,Libreville,0.38,9.45,Africa/Libreville,1715849100,2024-05-16 09:45,29.0,84.2,Partly cloudy,...,41.0,49.2,3,4,06:14,18:22,13:25,01:04,Waxing Gibbous,55
62,Gambia,Banjul,13.45,-16.58,Africa/Banjul,1715849100,2024-05-16 08:45,23.0,73.4,Sunny,...,12.9,59.6,1,2,06:40,19:26,14:01,02:04,Waxing Gibbous,55


In [206]:
# loc method 
df.loc[63,'air_quality_PM2.5']

np.float64(2.0)

### Renaming column

In [207]:
df.rename(columns={"location_name":"place"},inplace=True)
df

,country,place,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


In [208]:
df['condition_text'].value_counts()

condition_text
Partly cloudy                                  39376
Sunny                                          39229
Patchy rain nearby                             11376
Partly Cloudy                                   8669
Overcast                                        7410
Clear                                           7200
Mist                                            5716
Light rain                                      4287
Light rain shower                               3133
Fog                                             1709
Cloudy                                          1633
Moderate or heavy rain with thunder             1090
Moderate rain                                   1042
Patchy light rain with thunder                   843
Light drizzle                                    678
Light snow                                       523
Patchy light drizzle                             455
Moderate or heavy rain shower                    287
Thundery outbreaks in nearby   

In [209]:
# parse last_updated as datetime
df["last_updated"] = pd.to_datetime(df["last_updated"],errors="coerce")
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 41 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   country                       136633 non-null  str           
 1   place                         136633 non-null  str           
 2   latitude                      136633 non-null  float64       
 3   longitude                     136633 non-null  float64       
 4   timezone                      136633 non-null  str           
 5   last_updated_epoch            136633 non-null  int64         
 6   last_updated                  136633 non-null  datetime64[us]
 7   temperature_celsius           136633 non-null  float64       
 8   temperature_fahrenheit        136633 non-null  float64       
 9   condition_text                136633 non-null  str           
 10  wind_mph                      136633 non-null  float64       
 11  wind_kph                

### deleting columns

In [210]:
df.drop(columns=['last_updated_epoch'],inplace=True)
df


,country,place,latitude,longitude,timezone,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,wind_mph,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,2024-05-16 13:15:00,26.6,79.8,Partly Cloudy,8.3,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,2024-05-16 10:45:00,19.0,66.2,Partly cloudy,6.9,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,2024-05-16 09:45:00,23.0,73.4,Sunny,9.4,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,2024-05-16 10:45:00,6.3,43.3,Light drizzle,7.4,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,2024-05-16 09:45:00,26.0,78.8,Partly cloudy,8.1,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,2026-04-20 02:30:00,17.5,63.6,Partly Cloudy,2.2,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,2026-04-20 13:30:00,28.3,82.9,Cloudy,11.0,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,2026-04-20 09:30:00,21.0,69.7,Sunny,4.9,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,2026-04-20 08:30:00,21.5,70.8,Sunny,8.1,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


In [211]:
df.drop(columns=['temperature_fahrenheit','air_quality_PM2.5'],inplace=True)
df

,country,place,latitude,longitude,timezone,last_updated,temperature_celsius,condition_text,wind_mph,wind_kph,...,air_quality_Sulphur_dioxide,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,8.3,13.3,...,0.20,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,2024-05-16 10:45:00,19.0,Partly cloudy,6.9,11.2,...,0.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,2024-05-16 09:45:00,23.0,Sunny,9.4,15.1,...,13.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,2024-05-16 10:45:00,6.3,Light drizzle,7.4,11.9,...,0.20,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,2024-05-16 09:45:00,26.0,Partly cloudy,8.1,13.0,...,31.50,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,2026-04-20 02:30:00,17.5,Partly Cloudy,2.2,3.6,...,1.65,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,2026-04-20 13:30:00,28.3,Cloudy,11.0,17.6,...,21.05,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,2026-04-20 09:30:00,21.0,Sunny,4.9,7.9,...,2.05,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,2026-04-20 08:30:00,21.5,Sunny,8.1,13.0,...,2.05,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


In [212]:
df.isnull().sum()

country                         0
place                           0
latitude                        0
longitude                       0
timezone                        0
last_updated                    0
temperature_celsius             0
condition_text                  0
wind_mph                        0
wind_kph                        0
wind_degree                     0
wind_direction                  0
pressure_mb                     0
pressure_in                     0
precip_mm                       0
precip_in                       0
humidity                        0
cloud                           0
feels_like_celsius              0
feels_like_fahrenheit           0
visibility_km                   0
visibility_miles                0
uv_index                        0
gust_mph                        0
gust_kph                        0
air_quality_Carbon_Monoxide     0
air_quality_Ozone               0
air_quality_Nitrogen_dioxide    0
air_quality_Sulphur_dioxide     0
air_quality_PM

In [213]:
df['condition_text'].mode()

0    Partly cloudy
Name: condition_text, dtype: str

In [221]:
num_col = df[['latitude','longitude','temperature_celsius','wind_mph','wind_kph','air_quality_Sulphur_dioxide','air_quality_PM10','moon_illumination']]
for i in num_col:
   
   print(i,df[i].median())


latitude 17.25
longitude 23.2361
temperature_celsius 23.9
wind_mph 6.7
wind_kph 10.8
air_quality_Sulphur_dioxide 2.405
air_quality_PM10 19.98
moon_illumination 49.0
